internal/controllers/provisioningrequest_controller.go

```diff
@@ -386,81 +384,184 @@ func (t *provisioningRequestReconcilerTask) validateProvisioningRequestCR(ctx co
+	}
+
+	if err = t.validatePolicyTemplateInputMatchesSchema(ctx, clusterTemplate); err != nil {
+		return fmt.Errorf("failed to validate PolicyTemplate input: %w", err)
+	}
+
+	// TODO: Verify that ClusterInstance is per ClusterRequest basis.
+	//       There should not be multiple ClusterRequests for the same ClusterInstance.
+	return nil
+}
+
+// validateClusterInstanceInputMatchesSchema validates that the ClusterInstance input
+// from the ProvisioningRequest matches the schema defined in the ClusterTemplate.
+// If valid, the merged ClusterInstance data is stored in the clusterInput.
+func (t *provisioningRequestReconcilerTask) validateClusterInstanceInputMatchesSchema(
+	ctx context.Context, clusterTemplate *provisioningv1alpha1.ClusterTemplate) error {
+
+	// Get the subschema for ClusterInstanceParameters
+	clusterInstanceSubSchema, err := utils.ExtractSubSchema(
+		clusterTemplate.Spec.TemplateParameterSchema.Raw, utils.TemplateParamClusterInstance)
 	if err != nil {
-		return utils.NewInputError("failed to get the ClusterTemplate input for ClusterInstance: %s", err.Error())
+		return utils.NewInputError(
+			"failed to extract %s subschema: %s", utils.TemplateParamClusterInstance, err.Error())
 	}
-	var subSchema []byte
-	subSchema, err = utils.ExtractSubSchema(clusterTemplate.Spec.TemplateParameterSchema.Raw, clusterInstanceParametersString)
+	// Any unknown fields not defined in the schema will be disallowed
+	utils.DisallowUnknownFieldsInSchema(clusterInstanceSubSchema)
+
+	// Get the matching input for ClusterInstanceParameters
+	clusterInstanceMatchingInput, err := utils.ExtractMatchingInput(
+		t.object.Spec.TemplateParameters.Raw, utils.TemplateParamClusterInstance)
 	if err != nil {
-		return utils.NewInputError("failed to extract clusterInstanceParameters subSchema: %s", err.Error())
+		return utils.NewInputError(
+			"failed to extract matching input for subSchema %s: %w", utils.TemplateParamClusterInstance, err)
 	}
-	err = t.validateClusterTemplateInputMatchesSchema(
-		&runtime.RawExtension{Raw: subSchema},
-		clusterTemplateInputMap,
-		utils.ClusterInstanceDataType)
+	clusterInstanceMatchingInputMap := clusterInstanceMatchingInput.(map[string]any)
+
+	// Validate the ClusterInstance input from ProvisioningRequest matches the schema
+	err = utils.ValidateJsonAgainstJsonSchema(
+		clusterInstanceSubSchema, clusterInstanceMatchingInput)
 	if err != nil {
-		return utils.NewInputError("failed to validate ClusterTemplate input matches schema: %s", err.Error())
+		return utils.NewInputError(
+			"the provided %s does not match the schema from ClusterTemplate (%s): %w",
+			utils.TemplateParamClusterInstance, clusterTemplate.Name, err)
 	}
-	// Get the merged clusterinstance input data
-	mergedClusterInstanceData, err := t.getMergedClusterInputData(ctx, clusterTemplate, utils.ClusterInstanceDataType)
+
+	// Get the merged ClusterInstance input data
+	mergedClusterInstanceData, err := t.getMergedClusterInputData(
```

Comment:
Optional: Add a comment explaining why we cannot validate the merged template at this point, something like: "The clusterInstance template defined in clusterInstanceSubSchema is only a subset of the template of the SiteConfig Clustertemplate CR. However, the content of the cluster request "clusterInstanceDefaults" includes input matching the Siteconfig Full ClusterTemplate. So at this point we cannot do further validate as some parameters might exist in the Full Siteconfig template that are not in the ORAN template. The Full Cluster instance validation is performed in the Cluster Instance CR dry run"

Reply:
Updated